### DISCUSSION TITLE
passing the arg through stdin instead of as command line arg for `dfx canister call`

### DISCUSSION THREAD

**Discussion by luc-mercier** [dfinity/sdk#2656] [Feature Requests]
Sometimes passing the argument as commend line flag is not so convenient.

Two cases are:
* Canisters that take in raw binary data that is not Candid, and the caller wants to call some function to encode it. Using $(...) is not always easy. In the example of the registry canister, while I know I can encode in command line an argument, I never succeeded in wrapping it inside a $(...), the shell is doing something that I don't want.

* Reading from a file, when the argument is large. This applies to both raw binary and candid text representation.

I suggest that there is some flag to read the arg from stdin.

**Comment by chenyan-dfinity** [comment]
Just curious, which canister is not using Candid and why?

**Comment by luc-mercier** [comment]
The registry is not using Candid, for instance.

Why: I wasn't involved in the decision, I'm not sure.

But I'm pretty sure that as long as it's allowed to use arbitrary bytes for their interfaces, some developers will do so.

**Comment by nomeata** [comment]
A `--raw` flag is also useful to test Candid canisters with invalid or tricky Candid values, so this is reasonable.

`hfx call` supports `--raw-arg` as an alternative to `--candid`, but admittedly it is a more low-level tool.

**Comment by luc-mercier** [comment]
@nomeata I actually have the same FR for `hfx`: with `hfx call --raw-arg`, I'm also hitting the same problem: using `$(...)` to evaluate the `--raw-arg` causes difficulties with the shell. I've never managed to pass a binary proto to `hfx call` for instance, despite having confirmation that my command inside the `$(...)` does what it should.

**Comment by nomeata** [comment]
Right, reading from a file or stdin is a useful feature, but it’s kinda orthogonal to Candid vs. Raw, isn’t it?

**Comment by luc-mercier** [comment]
> Right, reading from a file or stdin is a useful feature, but it’s kinda orthogonal to Candid vs. Raw, isn’t it?

Oh, absolutely. It was just to give the motivation for this request.

**Comment by nomeata** [comment]
Is this a “nice to have” or a “this would make my life better on a regular basis” for you?

**Comment by luc-mercier** [comment]
I'm actually not sure.

I like to do manual testing of code I write in addition to automated testing. With e.g. the registry canister, I have NEVER succeeded in calling it by hand (purely because of this issue, I believe). However:
* automated testing works fine, I have added a e2e test for the registry, and that was much easier than trying to do manual testing
* for newer NNS canisters we've started using Candid, for which I should be able to pass the arg as text (not binary) and not have this issue.

So right now this PR is not a high priority for me. But I suspect the need will come again occasionally, maybe not during "pure coding" phases, but more in "let's make sure everything really works" or "debug stuff in prod" phases.

**Comment by nomeata** [comment]
Well, let us know if the pain becomes bigger!

**Comment by ghost** [comment]
This is now available through the `--argument-file` parameter, implemented in https://github.com/dfinity/sdk/pull/2328

Example:
```
$ echo '("XYZ")' | dfx canister call dd_backend greet --argument-file -
("Hello, XYZ!")
```
